tests/functional/test_application_restricted.py

```diff
@@ -0,0 +1,232 @@
+            "client_assertion": encoded_jwt,
+        },
+    )
+
+    response_json = response.json()
+
+    assert {"access_token", "expires_in", "token_type"} == response_json.keys()
+    assert response_json["access_token"] is not None
+    assert response_json["token_type"] == "Bearer"
+    try:
+        response_expires_in = int(response_json["expires_in"])
+        assert response_expires_in >= 0
+    except ValueError:
+        assert False, "Invalid 'expires_in' value. Must be an integer."
+
+    auth["response"] = response.json()
+
+
+@given("I have an invalid access token")
+def set_invalid_access_token(auth):
+    auth["access_token"] = "INVALID_ACCESS_TOKEN"
+
+
+@given("I have no access token")
+def set_no_access_token(auth):
+    auth["access_token"] = None
+
+
+@given("I have an expired access token")
+def set_expired_access_token(auth):
+    claims = {
+        "sub": config.APPLICATION_RESTRICTED_API_KEY,
+        "iss": config.APPLICATION_RESTRICTED_API_KEY,
+        "jti": str(uuid.uuid4()),
+        "aud": f"{config.BASE_URL}/oauth2/token",
+        "exp": int(time.time()),
+    }
+
+    headers = {
+        "kid": config.KEY_ID
+        }
+
+    encoded_jwt = jwt.encode(claims, config.SIGNING_KEY, algorithm="RS512", headers=headers)
+
+    response = requests.post(
+        f"{config.BASE_URL}/oauth2/token",
+        data={
+            "grant_type": "client_credentials",
+            "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+            "client_assertion": encoded_jwt,
+        }
+    )
+
+    print(response.json())
+
+    assert False
+
+
+@given("I have a request context", target_fixture="context")
+def context():
+    return {}
+
+
+@when("I GET a patient")
+def get_patient(auth, context):
+    authentication = auth["access_token"]
+
+    if authentication is not None:
+        token_type = auth["token_type"]
+        authentication = f"{token_type} {authentication}"
+
+    headers = {
+            "NHSD-SESSION-URID": "123",
+            "Authorization": f"Bearer {authentication}",
+        }
+
+    response = get_patient_request(headers)
+
+    context["response"] = response.json()
+    context["status"] = response.status_code
+
+
+@then(
+    parsers.cfparse(
+        "I get a {status:Number} HTTP response", extra_types=dict(Number=int)
+    )
+)
+def check_status(status, context):
+    print(context["response"])
+    assert context["status"] == status
+
+
+@then("I get a Patient resource in the response")
+def check_patient_resource(context):
+    expected_keys = {
+        "resourceType",
+        "timestamp",
+        "total",
+        "type"
+    }
+    assert context["response"].keys() == expected_keys
+    assert context["response"]["resourceType"] == "Bundle"
+    assert context["response"]["total"] == 0
+    assert context["response"]["type"] == "searchset"
+
+    try:
+        datetime.datetime.strptime(context["response"]["timestamp"], '%Y-%m-%dT%H:%M:%S+00:00')
```

Comment:
I've done this too in my tests, but used `dateutil.parser.parse`. Just informational -- I will likely switch it over when I do my merge